File: dp/fibo.py

def fibo(n, cache = {}) ->int:
        if n in cache:
            return cache[n]
        if n == 0: 
            return 0 
        
        elif n == 1:
            return 1

        cache[n] =  fibo(n-2, cache) + fibo(n-1, cache)
        return cache[n]

for i in range(10):
    print(i + 1, " ",fibo(i))

File: dp/fibo.java
import java.util.HashMap;

class fibo {

    public static void main(String[] args) {

        HashMap<Integer, Integer> memo = new HashMap<>();

        for (int i = 0; i < 10; i++) {
                System.out.println(fibo(i, memo));
        }
        
    }

    public static int fibo(int n, HashMap<Integer, Integer> memo) {
        
        if (n <= 1) return n; 

        if (memo.containsKey(n)) {
            return memo.get(n);
        }

        int result = fibo(n-1, memo) + fibo(n-2, memo);

        memo.put(n, result);

        return result;

    }
}
